Add script to load and tag new cqf ruler image

mgramigna opened the pull request:
# Summary

Add ability to load a fresh cqf-ruler with measure and patient data that is outputted from `fhir-bundle-calculator` (no valuesets).

## New behavior

Ability to run a script from the root directory and specify measure/docker version to tag

## Code changes

* Bash script to load cqf ruler with `MeasureReport` and patient data
* Root Makefile to run above script and tag/release new docker image.
* Added option to EXM105's makefile to use the measure report generation flag added in https://github.com/projecttacoma/fhir-bundle-calculator/pull/10

# Testing guidance

1) Run ye olde `make` of an individual measure's `Makefile` this will generate `output` in that measure's directory, which needs to be present for the script I added.
  * If #2 is not merged before this, then just use EXM105, as it is the only one whose files live here now
2) In the root directory of this repo, run `make MEASURE_DIR=/path/to/measure VERSION=x.y.z`
  * This will take the most recent output from 1), load a fresh cqf-ruler image with the patient and measure data, tag it with the version specified, and push it to dockerhub. 
  * **NOTE**: If you just want to test without pushing, just comment out line 19 of the `Makefile`

**Important**: The measure's Makefile now depends on https://github.com/projecttacoma/fhir-bundle-calculator/pull/10, as it expects a `MeasureReport` resource to be there. Options to get the most up to date are either set up an `npm link` with your `fhir-bundle-calculator` code on the branch `measurereport-generation` or replace `calculate-bundles` with `node /path/to/fhir-bundle/calculator/src/cli.js` while testing until the referenced PR is merged.

Godspeed.

Commit:
Add script to load and tag new cqf ruler image

lukoktonos commented on EXM105/Makefile:
```diff
@@ -21,6 +21,21 @@ connectathon:
+  http://localhost:8080/cqf-ruler-dstu3/fhir/Measure/measure-EXM105-FHIR3-8.0.000 \
+  -H 'Content-Type: application/json' \
+	-d @connectathon/fhir3/bundles/EXM105_FHIR3-8.0.000/EXM105_FHIR3-8.0.000-files/measure-EXM105_FHIR3-8.0.000.json
+
```
without `touch .seed-cqf-ruler-no-vs` this target will run even if was previously executed, which may result in duplicate resources added to the cqf-ruler.

mgramigna replied:
Resolved

lukoktonos commented on EXM105/Makefile:
```diff
@@ -21,6 +21,21 @@ connectathon:
 	-d @connectathon/fhir3/bundles/EXM105_FHIR3-8.0.000/EXM105_FHIR3-8.0.000-bundle.json
 
+.seed-cqf-ruler-no-vs:
+	until `curl --output /dev/null --silent --head --fail http://localhost:8080/cqf-ruler-dstu3`; do printf '.'; sleep 5; done
```
I think this wait should be moved to the top level Makefile before the `restart-cqf-ruler` target is considered complete

mgramigna replied:
I think I might defer on this one until the task that refactors all the Makefiles. As it is now, we still want to be able to run standalone `make`s from the individual measure directories, so I think in the meantime this might need to stay where it is.

Agree that it could be moved in the future thought.

lukoktonos commented on Makefile:
```diff
@@ -0,0 +1,24 @@
+info:
+	$(info usage: `make MEASURE_DIR=/path/to/measure/dir VERSION=x.y.z)
+
+.seed-cqf-ruler-no-vs:
```
Similar comment about creating a file to indicate that target has been "built" for the targets in this makefile

Target filenames should also be added to .gitignore and `rm`ed with the `clean` target

mgramigna replied:
Resolved

lukoktonos commented on Makefile:
```diff
@@ -0,0 +1,24 @@
+all: .restart-cqf-ruler .run-load-script .update-cqf-ruler-image
```
```suggestion
all: .restart-cqf-ruler .update-cqf-ruler-image
```

mgramigna replied:
Lumped this suggestion along with the other in one commit

lukoktonos commented on Makefile:
```diff
@@ -0,0 +1,24 @@
+.restart-cqf-ruler:
+	docker stop cqf-ruler || true && docker run --name cqf-ruler --rm -dit -p 8080:8080 contentgroup/cqf-ruler:develop
+
+.update-cqf-ruler-image:
```
```suggestion
.update-cqf-ruler-image: .seed-cqf-ruler-no-vs .run-load-script 
```

Commit:
Update makefile targets to cache, clean, and order properly

lukoktonos commented on Makefile:
```diff
@@ -1,11 +1,11 @@
 .seed-cqf-ruler-no-vs:
-	cd $(MEASURE_DIR)
-	make connectathon .seed-cqf-ruler-no-vs
+	make -C $(MEASURE_DIR) connectathon .seed-cqf-ruler-no-vs
```
If `MEASURE_DIR` isn't specified, this gives a slightly confusing error about the `connectathon` directory not existing.  Perhaps we should add a target that validates that that parameter has been defined.  Could also be part of @tstrass 's DRY out makefiles effort.

`.seed-cqf-ruler-no-vs` only exists for `EXM_105` currently, but since the other measures were moved over here, they will need to have that target added.  I think it makes sense to add them along with this change rather than putting it off for the DRY out in order to keep everything in a working state.

tstrass replied:
@mgramigna this can be part of https://jira.mitre.org/browse/TACOSTRAT-346 unless you'd rather address Luke's comment in this PR. LMK

mgramigna replied:
I'm addressing the second paragraph, but will probably rope the first paragraph about `MEASURE_DIR` into 346, yeah

lukoktonos commented on Makefile:
```diff
@@ -14,11 +14,12 @@ info:
 .restart-cqf-ruler:
 	docker stop cqf-ruler || true && docker run --name cqf-ruler --rm -dit -p 8080:8080 contentgroup/cqf-ruler:develop
```
```suggestion
	-docker stop cqf-ruler
	docker run --name cqf-ruler --rm -dit -p 8080:8080 contentgroup/cqf-ruler:develop
```

Adding a `-` to the beginning of a line in the makefile will prevent make from halting upon error and I think is a slightly cleaner way of doing this.  May want to add one to the second line as well.

Commit:
Add dash to prevent halting if container not already running

Co-Authored-By: Luke Osborne <luke.w.osborne@gmail.com>

lukoktonos commented:
Had a thought about the automatic tagging/pushing of docker images:

If we load the image without valuesets, `.seed-cqf-ruler-no-vs` will run.  Then if we go add some valuesets and re-do the `make` command, we may end up accidentally pushing an image with valuesets to docker hub.  So I think we should query cqf-ruler ValueSet and confirm that there are 0 value sets in the same target that does the tagging/pushing.  Something like `curl ... | grep "count = 0"` or similar.

Commit:
Update seeding, calculation, and clean targets for each makefile

Commit:
Only grab first output directory ordered by time

Commit:
Added script to check for valuesets and push to dockerhub

mgramigna commented:
Comments should be resolved. Updated the Makefiles of all the other measures, and added checking for presence of valuesets before tagging/pushing new image.

lukoktonos reviewed (approved):
Changes look good.  Was able to push image without valuesets and not able to push one with valuesets. 👍
